nifi-commons/nifi-security-utils/src/main/java/org/apache/nifi/security/xml/SafeXMLConfiguration.java

```diff
@@ -0,0 +1,154 @@
+    private static final String W3C_XML_SCHEMA =
+            "http://www.w3.org/2001/XMLSchema";
+
+    // These features are used to disable processing external entities in the DocumentBuilderFactory to protect against XXE attacks
+    private static final String DISALLOW_DOCTYPES = "http://apache.org/xml/features/disallow-doctype-decl";
+    private static final String ALLOW_EXTERNAL_GENERAL_ENTITIES = "http://xml.org/sax/features/external-general-entities";
+    private static final String ALLOW_EXTERNAL_PARAM_ENTITIES = "http://xml.org/sax/features/external-parameter-entities";
+    private static final String ALLOW_EXTERNAL_DTD = "http://apache.org/xml/features/nonvalidating/load-external-dtd";
+
+    private static final String XXE_ERROR_MESSAGE = "XML configuration file contained an external entity. To eliminate XXE vulnerabilities, NiFi has external entity processing disabled.";
+
+    @Override
+    public void initFileLocator(FileLocator loc) {
+        super.initFileLocator(loc);
+    }
+
+    public SafeXMLConfiguration() {
+        super();
+    }
+
+    public SafeXMLConfiguration(HierarchicalConfiguration<ImmutableNode> c) {
+        super(c);
+    }
+
+    /**
+     * This overridden createDocumentBuilder() method sets the appropriate factory attributes to disable XXE parsing.
+     * @return Returns a safe DocumentBuilder
+     * @throws ParserConfigurationException A configuration error
+     */
+    @Override
+    public DocumentBuilder createDocumentBuilder() throws ParserConfigurationException {
+        if (getDocumentBuilder() != null) {
+            return getDocumentBuilder();
+        }
+
+        final DocumentBuilderFactory factory = DocumentBuilderFactory
+                .newInstance();
+
+        if (isValidating()) {
+            factory.setValidating(true);
+            if (isSchemaValidation()) {
+                factory.setNamespaceAware(true);
+                factory.setAttribute(JAXP_SCHEMA_LANGUAGE, W3C_XML_SCHEMA);
+            }
+        }
+
+        // Disable DTDs and external entities to protect against XXE
+        factory.setAttribute(DISALLOW_DOCTYPES, true);
+        factory.setAttribute(ALLOW_EXTERNAL_GENERAL_ENTITIES, false);
+        factory.setAttribute(ALLOW_EXTERNAL_PARAM_ENTITIES, false);
+        factory.setAttribute(ALLOW_EXTERNAL_DTD, false);
+
+        final DocumentBuilder result;
+        result = factory.newDocumentBuilder();
+        result.setEntityResolver(super.getEntityResolver());
+
+        if (isValidating()) {
+            // register an error handler which detects validation errors
+            result.setErrorHandler(new DefaultHandler() {
+                @Override
+                public void error(SAXParseException ex) throws SAXException {
+                    throw ex;
+                }
+            });
+        }
+
+        return result;
+    }
+
+    @Override
+    public void read(Reader in) throws ConfigurationException, IOException {
+        try {
```

Comment:
I refactored this duplicate code into a private method (see [d67d6e4](https://github.com/alopresto/nifi/commit/d67d6e4f0935c6c880e6874663fd85f24e825707)).